### PR TITLE
WIP: introduce scheduling method Selector

### PR DESCRIPTION
**Pull request by mandelsoft**
**What this PR does / why we need it**:
The new scheduling method `Selector` allows to make use of the new label selector field
`providers` to select seeds running on different providers depending of the provider list list and matching labels. It prefers matching seeds if the load is the same.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
If automatic cross provider scheduling is desired, it  is possible know to specify a new seed selector field `providers` for a cloud profile to enable scheduling on seeds running on different providers.
It is observed if the scheduling method `Selector` has been chosen, which is the new default now.
```

### CODE REVIEW

**Review by danielfoehrKn — COMMENTED** [review]
as discussed with @mandelsoft 

The scheduling decision is now (compared to v1beta1) dependent on the CloudProfile (SeedSelector & ProviderSelector). So it would make sense to add an optional  SchedulingStrategy into the Cloudprofile that overrides the default configuration for the Scheduler. 

The Selector Strategy will determine candidates/seeds for each provider(or wildcard) that is defined in the cloudprofile SeedSelector.Providers. So a Shoot on gcp can be scheduled to a Seed on aws.
Currently  the Shoot will be scheduled to the Seed with the least amount of Shoots among all seed candidates. This is the only scheduling factor.
This might lead to scheduling decisions where a Seed from a different provider is being preferred over a Seed with the same provider, only because it has a slightly lower amount of Shoots.

Ideally we would have weight factors defining
- weight of being the same provider (as the Shoot)
- weight of being the same provider and same region
- weight of Shoot load 

We could also define these factors in the CloudProfile along the Scheduling Strategy.

Also we need integration test for the new SchedulingStrategy.

WDYT @rfranzke

**Comment by rfranzke** [comment]
Okay, so what you actually want to achieve is that we want to lift the restriction that the scheduler always ensures seed-provider = shoot-provider.

Well, TBH,

> So it would make sense to add an optional SchedulingStrategy into the Cloudprofile that overrides the default configuration for the Scheduler.

sounds odd. Why do we need this? Isn't it enough to do the following:

* The `CloudProfile` gets extended with  the new `providers[]` field which is defaulted to the `.spec.type`, e.g.

```yaml
apiVersion: core.gardener.cloud/v1alpha1
kind: CloudProfile
spec:
  type: aws
  seedSelector:
    matchLabels: {} # matches all labels
    providers: ["aws"]
```

* The gardener-scheduler gets adapted such that it behaves as following:
  * It finds all seed candidates of those providers respected by the `CloudProfile`'s `.spec.seedSelector.providers[]` field referenced by the `Shoot` **and** matching the `.spec.seedSelector.matchLabels`.
  * If the strategy is `SameRegion` it filters out all seeds (of the same provider) whose `.spec.provider.region` not equals to `.spec.region` of the shoot. If the strategy is `MinimalDistance` it keeps filtering out those which are "too far away" ([like today](https://github.com/gardener/gardener/blob/master/pkg/scheduler/controller/shoot/scheduler_control.go#L265-L274))
    * This means that all seeds (in all regions) of potentially other selected providers are considered as we cannot judge the region/distance criteria here.
  * Next, it filters out all seeds that are tainted as `invisible` or not ready.
  * Next, it filters out all seeds whose networks are not disjoint with the shoot's networks.
  * Now, it has a list of all potential candidates, and now you can implement weight functions that find the best one.
    * Weight factor 1: Seeds with the same provider like shoots get a higher weight.
    * Weight factor 2: Seeds with lower usage get a higher weight.

> This might lead to scheduling decisions where a Seed from a different provider is being preferred over a Seed with the same provider, only because it has a slightly lower amount of Shoots.

This cannot happen anymore now as we always prefer the same provider (weight factor 1). Only if no seed of the same provider exists then other providers are considered. And yes, I would not make this configurable in the cloud profile as I don't see a use case for this, do you have a concrete example? I don't see a reason why we need to maintain scheduling strategies or even weight criteria per cloud profile at this point in time.

On the other hand, I also think that we could achieve that seeds of other providers are considered by the scheduler without introducing `.spec.seedSelector.providers[]` at all. Yes, this has the "disadvantage" that the seeds must be labeled appropriately (e.g., with the provider types), but then they can get selected via the existing `.spec.seedSelector`. Whether this is a disadvantage or not is personal preference, I guess ;-) I would not consider it that way as it is implementing a [standard Kubernetes concept](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/). In this case, we would only have to adapt the scheduler slightly to prioritize the `seedSelector` and only then the same provider type. This saves us from a bunch of additional code and complexity that needs to be maintained.

Though, I could also live with `.spec.seedSelector.providers[]` if the majority considers this approach as more appropriate. However, what I would try to avoid now is the even further additional complexity that would come with overwriting scheduling strategies in the cloud profile unless we have a valid use case that requires us to do so.

/cc @vlerenc @timuthy @ialidzhikov @vpnachev @zanetworker - opinions?

**Comment by vlerenc** [comment]
> Currently the Shoot will be scheduled to the Seed with the least amount of Shoots among all seed candidates. This is the only scheduling factor. This might lead to scheduling decisions where a Seed from a different provider is being preferred over a Seed with the same provider, only because it has a slightly lower amount of Shoots.

Oh, is that really the case? I thought we had minimal distance based on the naming (not ideal, but that's what I thought we did - it requires the same infrastructure anyways), so don't we pick today the **same** infrastructure and the geographically **closest** region and only then pick the **least utilised** seed? Just picking the least used seems odd to me.

Maybe what we should do is to take a look at the [Kubernetes scheduler](https://kubernetes.io/docs/concepts/scheduling/kube-scheduler/#kube-scheduler-implementation). It runs in two phases:
1. Filtering
2. Scoring

**Comment by rfranzke** [comment]
This depends on the configured strategy.. the default is that only seeds in the same region like the shoot are considered. However, we also have the `MinimalDistance` strategy where lexicographically "close" seeds are considered.
Only in the next step, when all the candidates have been determined, we are choosing the least utilised seed.

> Maybe what we should do is to take a look at the Kubernetes scheduler. It runs in two phases:
> 1. Filtering
> 1. Scoring

We already have this. In a first phase we filter out seeds of different providers, unavailable seeds, invisible seeds, seeds in different regions (see above), etc. In the next phase we score them (only one criteria today: "least utilised"), and then we pick the one with the best score...

File: pkg/apis/core/v1alpha1/types_cloudprofile.go [hunk 1]
```diff
@@ -67,14 +67,24 @@ type CloudProfileSpec struct {
 	// An empty list means that all seeds of the same provider type are supported.
 	// This is useful for environments that are of the same type (like openstack) but may have different "instances"/landscapes.
 	// +optional
-	SeedSelector *metav1.LabelSelector `json:"seedSelector,omitempty"`
+	SeedSelector *SeedSelector `json:"seedSelector,omitempty"`
```
Comment (vpnachev) [inline pkg/apis/core/v1alpha1/types_cloudprofile.go:70]: fyi, changing the type is not backward compatible change

File: pkg/apis/core/v1alpha1/types_cloudprofile.go [hunk 2]
```diff
@@ -67,14 +67,24 @@ type CloudProfileSpec struct {
...
 	// Type is the name of the provider.
 	Type string `json:"type"`
 	// VolumeTypes contains constraints regarding allowed values for volume types in the 'workers' block in the Shoot specification.
 	// +optional
 	VolumeTypes []VolumeType `json:"volumeTypes,omitempty"`
 }
 
+// SeedSelector contains constraints for selecting seed to be usable for shoots using a profile
```
Comment (vpnachev) [inline pkg/apis/core/v1alpha1/types_cloudprofile.go:78]: I do not understand this docu.

File: pkg/apis/garden/types.go [hunk 3]
```diff
@@ -82,13 +82,23 @@ type CloudProfileSpec struct {
...
 	// Type is the name of the provider.
 	Type string
 	// VolumeTypes contains constraints regarding allowed values for volume types in the 'workers' block in the Shoot specification.
 	VolumeTypes []VolumeType
 }
 
+// SeedSelector contains constraints for selecting seed to be usable for shoots using a profile
+type SeedSelector struct {
+	// LabelSelector is optional and can be used to select seeds by their label settings
+	// +optional
```
Comment (vpnachev) [inline pkg/apis/garden/types.go:95]:
```suggestion
```

File: pkg/apis/garden/types.go [hunk 4]
```diff
@@ -82,13 +82,23 @@ type CloudProfileSpec struct {
...
+	*metav1.LabelSelector
+	// Providers is optional and can be used by restricting seeds by their provider type. '*' can be used to enable seeds regardless of their provider type.
+	// +optional
```
Comment (vpnachev) [inline pkg/apis/garden/types.go:98]:
```suggestion
```

File: pkg/scheduler/apis/config/types.go [hunk 5]
```diff
@@ -22,12 +22,14 @@ import (
...
 const (
+	// Selector Strategy determines a seed candidate for a shoot base on SeedSelector preferring matching provider and region
+	Selector CandidateDeterminationStrategy = "Selector"
 	// SameRegion Strategy determines a seed candidate for a shoot only if the cloud profile and region are identical
 	SameRegion CandidateDeterminationStrategy = "SameRegion"
 	// MinimalDistance Strategy determines a seed candidate for a shoot if the cloud profile are identical. Then chooses the seed with the minimal distance to the shoot.
 	MinimalDistance CandidateDeterminationStrategy = "MinimalDistance"
 	// Default Strategy is the default strategy to use when there is no configuration provided
-	Default CandidateDeterminationStrategy = SameRegion
+	Default CandidateDeterminationStrategy = Selector
```
Comment (vpnachev) [inline pkg/scheduler/apis/config/types.go:32]: Changing default is not recommended without change the api version.

**Review by vpnachev — CHANGES_REQUESTED** [review]
Thanks!
You might find useful this [document](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md) about API conventions.

File: pkg/apis/core/v1alpha1/types_cloudprofile.go (hunk 1 above)
Comment (mandelsoft) [reply pkg/apis/core/v1alpha1/types_cloudprofile.go:70]: The `SeedSelector` is using the old `LabelSelector` as flat field. Therefore the new structure is an extension to the old one. The old manifests can be read by the new lib version and the new one can be read by the old one, potentially loosing the values for the additional fields, that are not used so far.

**Comment by rfranzke** [comment]
Can you please reopen the PR once the implementation + tests are ready? Thanks.

**Comment by mandelsoft** [comment]
We decided to redesign the complete scheduling mechanism to corrects handle field selectors and label selectors in combination with scheduling policies assigned to cloud profiles.
The policies should be able to handle cross platform assignments correctly, especially we need a metric to handle the nearest seed feature in all scenarios, including complete on-prem scenarios based on openstack and vmware using completely arbitrary region names. 
So the actual proposal is obsolete.

**Pull request closed without merging**